provider/pkg/clients/unstructured.go

```diff
@@ -18,43 +18,30 @@ import (
 	"fmt"
 
 	"github.com/pulumi/pulumi-kubernetes/provider/v3/pkg/kinds"
-	appsv1 "k8s.io/api/apps/v1"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
 	batchv1 "k8s.io/api/batch/v1"
 	corev1 "k8s.io/api/core/v1"
-	networkingv1 "k8s.io/api/networking/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
+	kscheme "k8s.io/client-go/kubernetes/scheme"
 )
 
-func FromUnstructured(obj *unstructured.Unstructured) (metav1.Object, error) {
-	var output metav1.Object
-	switch kinds.Kind(obj.GetKind()) {
-	case kinds.Deployment:
-		output = new(appsv1.Deployment)
-	case kinds.Job:
-		output = new(batchv1.Job)
-	case kinds.Ingress:
-		output = new(networkingv1.Ingress)
-	case kinds.PersistentVolume:
-		output = new(corev1.PersistentVolume)
-	case kinds.PersistentVolumeClaim:
-		output = new(corev1.PersistentVolumeClaim)
-	case kinds.Pod:
-		output = new(corev1.Pod)
-	case kinds.ReplicaSet:
-		output = new(appsv1.ReplicaSet)
-	case kinds.StatefulSet:
-		output = new(appsv1.StatefulSet)
-	default:
-		return nil, fmt.Errorf("unhandled Kind: %s", obj.GetKind())
+var scheme = runtime.NewScheme()
+
+// FromUnstructured dynamically converts an Unstructured Kubernetes resource into the typed equivalent. Only built-in
+// resource types are supported, so CustomResources will fail conversion with an error.
+func FromUnstructured(uns *unstructured.Unstructured) (runtime.Object, error) {
```

Comment:
I believe we'd always want our object to have the `.metadata` fields? If so, it might be better if we assert the union interface `client.Object` from [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/client/object.go) rather than just `runtime.Object` as the return type to prevent losing data.

So something like:

```
import "sigs.k8s.io/controller-runtime/pkg/client"

uObj, ok := obj.(client.Object)
if !ok { ... }
err = runtime.DefaultUnstructuredConverter.FromUnstructured(uns.Object, uObj)
```

Reply:
Sounds good!

Reply:
Actually, it looks like "sigs.k8s.io/controller-runtime/pkg/client" would be a new dependency for the provider. I'm not sure it's worth the potential hassle with upgrading dependencies for that.

I think `metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"` works for that as well.